Adding flags to optionally disable build unnecessary dependencies

recursion-ninja opened the pull request:
Addresses issues #25 .

RyanGlScott commented:
Thanks, @recursion-ninja.

I'd like to test this in the CI to ensure that this doesn't regress in the future. Can you make use of [`haskell-ci`](https://github.com/haskell-CI/haskell-ci/)'s `constraint-set` feature ([examples](https://github.com/ekmett/semigroupoids/blob/9f2b08f5eeacc2808d85aad1322eb221e0988702/cabal.haskell-ci#L8-L15)) so that the CI disables one of these flags per constraint set?

recursion-ninja commented:
@RyanGlScott Thanks for the feedback on the PR!

I actually wanted to do what you suggested but did not know the best way to go about it. Your direction to use `haskell-ci` was very valuable, as (at least for this use case), it is very easy to specify and perform these tests. I will likely use this tool again i the future for my own projects.

RyanGlScott commented:
Do note that `haskell-ci` won't automatically pick up your changes to the `cabal.haskell-ci` file—you'll need to run `haskell-ci regenerate` (after installing `haskell-ci`) in your `keys` checkout to actually regenerate the necessary CI YAML file.

recursion-ninja commented:
@RyanGlScott Again, thanks for the valuable guidance. The `haskell-ci` tool looks exactly like what I've wanted for a while, and tried to hand roll unsuccessfully.

I updated the `.github/workflows/haskell-ci.yml` via invoking `haskell-ci regenerate`.

The CI process appears to have incorporated new Cabal flags as well as successfully built all supported GHC versions.

I squashed all the commits together into one commit for this PR. I think the PR is ready to be merged.

RyanGlScott commented on keys.cabal:
```diff
@@ -65,7 +105,28 @@ library
+  if impl(ghc >= 9.6)
+    build-depends: transformers >= 0.6.1.0
+
+  if !impl(ghc >= 9.6) && impl(ghc > 7.10)
```
Minor nitpick: can you change `impl(ghc > 7.10)` to `impl(ghc >= 8.0)`? I find it easier to read the latter, and it's more consistent with the existing version bound conventions.

RyanGlScott commented on src/Data/Key.hs:
```diff
@@ -59,68 +63,110 @@ module Data.Key (
 import Control.Monad.Trans.Reader
 import qualified Data.Array as Array
 import Data.Array (Array)
+#if __GLASGOW_HASKELL__ > 710
```
Similarly, use `__GLASGOW_HASKELL__ >= 800` here (and everywhere else the file currently uses `__GLASGOW_HASKELL__ > 710`).

RyanGlScott commented on src/Data/Key.hs:
```diff
@@ -59,68 +63,110 @@ module Data.Key (
 import Control.Monad.Trans.Reader
 import qualified Data.Array as Array
 import Data.Array (Array)
+#if __GLASGOW_HASKELL__ > 710
+import Data.Foldable1 (Foldable1(..))
+#elif defined(MIN_VERSION_semigroupoids)
+import Data.Semigroup.Foldable (Foldable1(..))
+#endif
```
Speaking of which, why _does_ this code need to have CPP that's guarded against GHC 8.0? At a glance, I'm not sure what `Foldable1` has anything to do with GHC 8.0.

RyanGlScott commented:
Note that I recently pushed https://github.com/ekmett/keys/commit/af8c6aa4e80b0a75b74762e29722a78ab0811916, which requires `keys` to build with GHC 8.0 as the minimum. As such, I suspect that much of the CPP in this PR could be removed once rebased.
